internal/common/function/function.go

```diff
@@ -77,66 +77,125 @@ func ValidateBackupSpec(nonAdminBackup *nacv1alpha1.NonAdminBackup) error {
+	// Generate UUID suffix
+	uuidSuffix := uuid.New().String()
+
+	// Build the initial name based on the presence of namespace and nacName
+	nacObjectName := uuidSuffix
+	if len(nacName) > 0 {
+		nacObjectName = nacName + constant.NameDelimiter + nacObjectName
+	}
+	if len(namespace) > 0 {
+		nacObjectName = namespace + constant.NameDelimiter + nacObjectName
+	}
+
+	if len(nacObjectName) > constant.MaximumNacObjectNameLength {
+		// Calculate remaining length after UUID
+		remainingLength := constant.MaximumNacObjectNameLength - len(uuidSuffix)
+
+		delimeterLength := len(constant.NameDelimiter)
+
+		// Subtract two delimiter lengths to avoid a corner case where the namespace
+		// and delimiters leave no space for any part of nabName
+		if len(namespace) > remainingLength-delimeterLength-delimeterLength {
+			namespace = namespace[:remainingLength-delimeterLength-delimeterLength]
+			nacObjectName = namespace + constant.NameDelimiter + uuidSuffix
+		} else {
+			remainingLength = remainingLength - len(namespace) - delimeterLength - delimeterLength
+			nacName = nacName[:remainingLength]
+			nacObjectName = uuidSuffix
+			if len(nacName) > 0 {
+				nacObjectName = nacName + constant.NameDelimiter + nacObjectName
+			}
+			if len(namespace) > 0 {
+				nacObjectName = namespace + constant.NameDelimiter + nacObjectName
+			}
+		}
 	}
 
-	const hashLength = 14
-	nameHash := hex.EncodeToString(hasher.Sum(nil))[:hashLength] // Take first 14 chars
+	return nacObjectName
+}
 
-	usedLength := hashLength + len(constant.VeleroBackupNamePrefix) + len("--")
-	maxNamespaceLength := validation.DNS1123SubdomainMaxLength - usedLength
-	// Ensure the name is within the character limit
-	if len(namespace) > maxNamespaceLength {
-		// Truncate the namespace if necessary
-		return fmt.Sprintf("%s-%s-%s", constant.VeleroBackupNamePrefix, namespace[:maxNamespaceLength], nameHash)
+// GetObjectByLabel retrieves a list of Kubernetes objects of a specified type based on a label within a given namespace.
+// It returns a slice of the specified object type and an error.
+func GetObjectByLabel(ctx context.Context, clientInstance client.Client, namespace string, labelKey string, labelValue string, objectList client.ObjectList) error {
+	// Validate input parameters
+	if namespace == constant.EmptyString || labelKey == constant.EmptyString || labelValue == constant.EmptyString {
+		return fmt.Errorf("invalid input: namespace, labelKey, and labelValue must not be empty")
 	}
 
-	return fmt.Sprintf("%s-%s-%s", constant.VeleroBackupNamePrefix, namespace, nameHash)
+	labelSelector := labels.SelectorFromSet(labels.Set{labelKey: labelValue})
+
+	// Attempt to list objects with the specified label
+	if err := clientInstance.List(ctx, objectList, &client.ListOptions{
+		LabelSelector: labelSelector,
+		Namespace:     namespace,
+	}); err != nil {
+		return fmt.Errorf("failed to list objects in namespace '%s': %w", namespace, err)
+	}
+
+	return nil
```

Comment:
I would call this function something like `GetSingleObjectByLabel` and do validation here 

For case no items are found, I would use `NewNotFound` function https://pkg.go.dev/k8s.io/apimachinery@v0.29.0/pkg/api/errors#NewNotFound to keep behavior in controllers

Reply:
No, this is not how it works. This particular function is generic one that returns all the objects within that label. Later the function `GetVeleroBackupByLabel` is consuming that generic function and have single object logic. I don't want to call it Single, because it's single by name and not the `GetVeleroBackupsByLabel` which would be plural.

Reply:
I've moved this function to be more generic for use in restore as well in the future.

Reply:
I don't want to use proposed function, because we do not have knowledge *why* the error when listing the objects was raised. It's an error from the `List()` function, so there may be multiple reasons why it failed to List and even if the objects are in the cluster the error still may happen.

Reply:
Yeah lets cascade the list err as is for transparency.

Reply:
@shubham-pampattiwar by saying lets cascade you agree current implementation is correct or it needs changes?

I've implemented the function `GetObjectByLabel` to do just it's job, so no logic involved in the body of that function. It's currently used by:

https://github.com/migtools/oadp-non-admin/blob/97dacc63f1f009c9b6422b02165438614d482a29/internal/common/function/function.go#L150-L161

And the above `GetVeleroBackupByLabel` does cascade errors and lists accordingly.

We do not know what other parts such as sync controller or NAR controller will do with the situation where no objects based on the label value are found so it's up to the consumer function to decide how to handle such situation. I believe current implementation is generic and correct, because it separates `get me some object X` from logical function that handles the result for the X.

It would be different implementation if that function was not part of our `common` functions and specific **only** to the NAB object.

Reply:
Current implementation looks good, just update the func name to reflect the action.

Reply:
Updated generic function name to:

`ListLabeledObjectsInNamespace`

However reading the code again I think the `GetVeleroBackupByLabel` is correct. It only returns object when exactly one is found, so it does not list velero backups by label.